Multithreading 02

engineer027 opened the pull request.

engineer027 committed:
added solution

engineer027 committed:
fix checkstyle

engineer027 committed:
add test

Sofasmile commented on src/main/java/multithreading/Main.java:
```diff
@@ -1,11 +1,22 @@
+        System.out.println("forkJoin " + forkJoin.compute());
+    }
+
+    private static List<Integer> getGenerateList(int size) {
```
давай винесемо цей метод в окремий клас який буде займатись генерацією списків з числами. Клас можна назвати: ListOfNumbersCreator or ListOfNumbersGenerator

Sofasmile commented on src/main/java/multithreading/MyCallableImpl.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class MyCallableImpl implements Callable<Integer> {
```
не дуже круто, коли в нас Callable і ExecutorService знаходяться в одному класі
краще мати окремий клас який наслідується від Callable і приймає на вхід список з числами, і в методі call() просто рахує їх суму і повертає результат.
Тому пропоную розділити цей клас на 2: SumCalculatorCallable і наприклад ExecutorServiceCalculator

Sofasmile commented on src/main/java/multithreading/MyCallableImpl.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public Integer getSum() {
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
```
давай тут будемо використовувати метод newFixedThreadPool для створення пула потоків

Sofasmile commented on src/main/java/multithreading/MyCallableImpl.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.concurrent.Future;
+
+public class MyCallableImpl implements Callable<Integer> {
+    private static final int SUB_LIST_SIZE = 10;
```
це заналто малий розмір підлістів, пропоную збільшити його хоча б до 100

Sofasmile commented on src/main/java/multithreading/MyForkJoinImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+public class MyForkJoinImpl extends RecursiveTask<Integer> {
+    private static final int SUB_LIST_SIZE = 10;
+    private static final int THRESHOLD = 1000000;
```
чому тут THRESHOLD рівний розміру списку з числами?

Sofasmile commented on src/main/java/multithreading/MyForkJoinImpl.java:
```diff
@@ -0,0 +1,43 @@
+            return ForkJoinTask.invokeAll(createSubtasks(list)).stream()
+                    .mapToInt(ForkJoinTask::join)
+                    .sum();
+        } else {
+            return processing();
+        }
```
варто спростити конструкцію

Sofasmile commented on src/test/java/multithreading/MyCallableImplTest.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.*;
```
використання автоімпортів погана практика

Sofasmile commented on src/test/java/multithreading/MyCallableImplTest.java:
```diff
@@ -0,0 +1,61 @@
+
+public class MyCallableImplTest {
+
+    List<Integer> list = new ArrayList<>(5);
```
не забуваємо про модифікатори
а також це можна винести в метод @ Before

Sofasmile commented on src/test/java/multithreading/MyCallableImplTest.java:
```diff
@@ -0,0 +1,61 @@
+        MyForkJoinImpl forkJoin = new MyForkJoinImpl(list);
+        Integer compute = forkJoin.compute();
+        Assert.assertEquals(20, compute.intValue());
+        list.clear();
```
'list.clear()' можна винести в метод After

engineer027 committed:
- create ExecutorServiceCalculator class;
- create ListOfNumbersCreator class;
- create SumCalculatorCallable class;
- update test

Sofasmile commented on src/test/java/multithreading/ExecutorServiceCalculatorTest.java:
```diff
@@ -0,0 +1,70 @@
+        list.clear();
+    }
+
+}
```
варто додати пусту строку
